### PR TITLE
Code for Admin Module (User Management)

### PR DESCRIPTION
**Pull request by chirag-compro**

### CODE REVIEW

File: client/app/_services/user.service.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+export class UserService {
+public userData: any;
+public currentUser: any;
+public useradded: boolean;
+
+ constructor(private http:HttpClient ) { 
+  }
+  getUsers(): Observable<User[]>{
+    if(this.userData && this.useradded == false){
+      return Observable.of(this.userData);
```
Comment (ashish-compro) [inline client/app/_services/user.service.ts]: It should call the server to get the updated data always when the component of user list is displayed.

File: client/app/_services/user.service.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import { HttpClient } from './http.client';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/observable/of';
+
+import { User } from './userModel';
+@Injectable()
+export class UserService {
+public userData: any;
```
Comment (ashish-compro) [inline client/app/_services/user.service.ts]: According to me this data should either be moved to a data store or should be fetched from the server every time.

File: client/app/_services/user.service.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  getUsers(): Observable<User[]>{
+    return this.http.get("/api/user")
+                    .map(this.extractData.bind(this))
+                    .catch(this.handleError);
+  }
+  private extractData(res: Response) {
+    let body = res.json();
+    return body || { };
+  }
+  private handleError (error: Response | any) {
```
Comment (chirag-compro) [inline client/app/_services/user.service.ts:25]: error handler can be a common service for all the modules.

File: client/app/administration/user-details-form/user-details-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                .subscribe(result => {
+                        this.message = result.message;
+                });
+      }
+    }
+    else return;
+  }
+
+  ngOnInit() {
+    this.instance = jQuery('.parsleyjs').parsley();
```
Comment (chirag-compro) [inline client/app/administration/user-details-form/user-details-form.component.ts]: rename instance  to ParsleyForm

File: client/app/administration/user-list/user-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      .focusout((e) => {
+      jQuery(e.target).closest('.input-group').removeClass('focus');
+    });
+  }
+  initializeUserData(): void{
+    this.route.data
+                     .subscribe(
+                       Data => this.Data = Data,
+                       error =>  this.errorMessage = <any>error);
+					   this.userData = this.Data["userData"];
```
Comment (chirag-compro) [inline client/app/administration/user-list/user-list.component.ts]: this.Data = Data, can be replaced with this.userData = this.Data["userData"];

File: client/app/administration/add-user/add-user.component.html [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<h1 class="page-title">Add <span class="fw-semi-bold">User</span></h1>
+<section class="widget" widget>
+  <div class="row">
+    <div class="col-lg-6 offset-lg-3 col-xs-12">
+      <app-user-details-form [title]="'Add the information about the user you want to add in Sims Builder.'" [mode]="'add'" (closeFormEvent)="closeFormEventListener($event)"></app-user-details-form>
```
Comment (chirag-compro) [inline client/app/administration/add-user/add-user.component.html]: closeFormEventListener can be renamed to closeFormListener

File: client/app/administration/user-details-form/user-details-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+@Component({
+  selector: 'app-user-details-form',
+  templateUrl: './user-details-form.component.html',
+  styleUrls: ['./user-details-form.component.scss']
+})
+export class UserDetailsFormComponent implements OnInit, OnChanges {
+@Input() title:string;
+@Input() user;
+@Input() mode;
+@Output() closeFormEvent: EventEmitter<any> = new EventEmitter();
```
Comment (chirag-compro) [inline client/app/administration/user-details-form/user-details-form.component.ts]: decide a better name for this event closeFormEvent and send message with each emit

File: client/app/administration/add-user/add-user.component.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  styleUrls: ['./add-user.component.scss']
+})
+export class AddUserComponent implements OnInit {
+  constructor() { }
+  ngOnInit() {
+     Messenger.options = { extraClasses: 'messenger-fixed messenger-on-top',
+							theme: 'block'}
+   }
+  closeFormEventListener(message){
+    if(message){
```
Comment (chirag-compro) [inline client/app/administration/add-user/add-user.component.ts]: Map of message emited and message shown can be implemented.

File: server/routes/index.js [hunk 9]
```diff
@@ -4,13 +4,14 @@ var loginRouter = require('./login/login.route');
 var taskBuilderData = require('./taskBuilderData');
 var skillRoutes = require('./skill.routes');
 var verifyToken = require('./login/verifyToken');
-
+var userData = require('./userData');
```
Comment (tanujaggarwal) [inline server/routes/index.js]: rename to user

File: server/routes/userData.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
+const express = require('express');
+const router = express.Router();
+const fs = require('fs');
```
Comment (tanujaggarwal) [inline server/routes/userData.js:3]: delete this line

File: server/routes/userData.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		res.json({
+				success: successFlag,
+				message: messageToSend,
+			});
+	})
+})
+router.put("/",function(req,res){
+	var user = req.body;
+	var userUpdateData = {  $set:{
```
Comment (tanujaggarwal) [inline server/routes/userData.js:65]: we can set parameters if only present in the request
request should contain the parameters modified only.

File: server/routes/index.js [hunk 12]
```diff
@@ -4,13 +4,14 @@ var loginRouter = require('./login/login.route');
...
 module.exports = function () {
 	router.use('/login', loginRouter());
 
 	router.use('*', verifyToken);
 	router.use('/skill', skillRoutes);
   router.use('/fetchTaskData', taskBuilderData);	
+	router.use('/user', userData);
```
Comment (tanujaggarwal) [inline server/routes/index.js]: If all users should be allowed to access this API or only admin users should be allowed
Baban sir to confirm.

File: client/app/administration/add-user/add-user.component.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  styleUrls: ['./add-user.component.scss']
+})
+export class AddUserComponent implements OnInit {
+  constructor() { }
+  ngOnInit() {
+     Messenger.options = { extraClasses: 'messenger-fixed messenger-on-top',
+							theme: 'block'}
+   }
+  emittedEventListener(message){
+    if(Map[message]!= "No Changes"){
```
Comment (vsingb3) [inline client/app/administration/add-user/add-user.component.ts]: All the valid messages should be declared in an enum

File: client/app/home/navbar/navbar.component.ts [hunk 14]
```diff
@@ -31,11 +38,27 @@ export class Navbar implements OnInit {
     this.authenticationService.logout();
     this.router.navigate(['/login']);
   }
-
+  checkUsertype(userRole){
```
Comment (vsingb3) [inline client/app/home/navbar/navbar.component.ts]: IsAdminUser

**Review by vsingb3 — APPROVED** [review]

**Pull request merged**